ament_copyright/ament_copyright/__init__.py

```diff
@@ -12,31 +12,354 @@
+            len(copyright_year)
+        if copyright_year[-4:] != str(int(current_year) - 1):
+            # append current year
+            content = content[:end_index] + ', ' + current_year + \
+                content[end_index:]
+        elif copyright_year[-5:-4] == '-':
+            # update end year of range
+            content = content[:end_index - 4] + current_year + \
+                content[end_index:]
+        else:
+            # update year to be a range
+            content = content[:end_index] + '-' + current_year + \
+                content[end_index:]
+
+        # output beginning of file for debugging
+        # index = end_index
+        # for _ in range(3):
+        #     index = get_index_of_next_line(content, index)
+        # print('<<<')
+        # print(content[:index - 1])
+        # print('>>>')
+
+        with open(filename, 'w') as h:
+            h.write(content)
+
+
+def add_copyright(filename, copyright):
+    """
+    Add the copyright message to a file.
+
+    The copyright is placed below an optional shebang line as well as an
+    optional coding line.
+    It is preceded by an empty line if not at the beginning of the file and
+    always followed by an empty line.
+    """
+    with open(filename, 'r') as h:
+        content = h.read()
+
+    begin_index = scan_past_coding_and_shebang_lines(content)
+    end_index = scan_past_empty_lines(content, begin_index)
+
+    # inject copyright message
+    comment = get_comment(filename, copyright)
+    inserted_block = '%s\n\n' % comment
+    if begin_index > 0:
+        inserted_block = '\n' + inserted_block
+    content = content[:begin_index] + inserted_block + content[end_index:]
+
+    # output beginning of file for debugging
+    # index = end_index + len(inserted_block)
+    # for _ in range(3):
+    #     index = get_index_of_next_line(content, index)
+    # print('<<<')
+    # print(content[:index - 1])
+    # print('>>>')
+
+    with open(filename, 'w') as h:
+        h.write(content)
+
+
+def scan_past_coding_and_shebang_lines(content):
+    index = 0
+    while (
+        is_comment_line(content, index) and
+        (is_coding_line(content, index) or
+         is_shebang_line(content, index))
+    ):
+        index = get_index_of_next_line(content, index)
+    return index
+
+
+def scan_past_empty_lines(content, index):
+    while is_empty_line(content, index):
+        index = get_index_of_next_line(content, index)
+    return index
+
+
+def get_index_of_next_line(content, index):
+    index_n = content.find('\n', index)
+    index_r = content.find('\r', index)
+    if index_n == -1 and index_r == -1:
+        return len(content)
+    if index_n == -1:
+        return index_r + 1
+    if index_r == -1:
+        return index_n + 1
+    return min(index_n + 1, index_r + 1)
+
+
+def is_comment_line(content, index):
+    return content[index] == '#' or content[index:index + 1] == '//'
```

Comment:
In order to handle block comments the whole logic would need to be modified (and will become more complex since `is_comment_line()` requires state or must operate on multiple lines). Therefore I would only aim to support `//` as well as `#` for the copyright message.

This should be described in the docs (part of ros2/ros2#3).